PHP 8.2 | WP_UnitTestCase_Base: explicitly set dynamic property

jrfnl opened the pull request:
Dynamic (non-explicitly declared) properties are deprecated as of PHP 8.2 and are expected to become a fatal error in PHP 9.0.

In this particular case, the `scandir()` method sets a dynamic property, which is subsequently used in the `delete_folders()` method.
Explicitly declaring the property fixes the PHP 8.2 incompatibility.

---

Having said that.... IMO this is not the right solution, but solving this properly could be considered a breaking change as these methods are both `public` and in the abstract base class for all tests.

Note: within the WP Core test suite, these methods are _only_ used in the `Tests_Multisite_GetSpaceUsed` test class.

The "proper" solution, IMO, would be as follows:
1. Change the `scandir()` method to return an array of the found directories. The `delete_folders()` method then would just need to use the return value and would not need access to the property either.
    I imagine this was previously not done due to the person writing the code being uncomfortable with recursive functions, but it is perfectly possible to do so and would get rid of the unnecessary property.
    The method is currently a `void` method and sets the property.
2. Move both methods to the `Tests_Multisite_GetSpaceUsed` test class as clearly they are not needed elsewhere.

Even just implementing step 1 would already be a good improvement and would still leave these methods available to future and/or external integration tests.

I'm happy to convert this PR to the proper solution, but would like a second opinion first. /cc @johnbillion 

---

Added a second commit:

### PHP 8.2 | WP_UnitTestCase_Base: minor refactor to remove need for $matched_dirs property

Follow-up to the earlier commit which explicitly declared the `$matched_dirs` property.

This commit implements the additional recommendation to remove the need for the - previously dynamic - property completely.

Effectively, this changes the `WP_UnitTestCase_Base::scandir()` method to return an array with matched directories instead of setting the property by using recursion in the method itself in an optimized manner.

Note the `array_merge()` not being in the loop itself, but at the very end of the function. This is for performance reasons.
See https://github.com/dseguy/clearPHP/blob/master/rules/no-array_merge-in-loop.md for a more detailed explanation of this.

I have verified in detail that the actual results of the previous version of this method and this version match, even when the paths passed are more complex and have deeper nested subdirectories.

---


Trac ticket: https://core.trac.wordpress.org/ticket/56033

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

jrfnl committed:
PHP 8.2 | WP_UnitTestCase_Base: explicitly set dynamic property

Dynamic (non-explicitly declared) properties are deprecated as of PHP 8.2 and are expected to become a fatal error in PHP 9.0.

In this particular case, the `scandir()` method sets a dynamic property, which is subsequently used in the `delete_folders()` method.
Explicitly declaring the property fixes the PHP 8.2 incompatibility.

---

Having said that.... IMO this is not the right solution, but solving this properly could be considered a breaking change as these methods are both `public` and in the abstract base class for all tests.

Note: within the WP Core test suite, these methods are _only_ used in the `Tests_Multisite_GetSpaceUsed` test class.

The "proper" solution, IMO, would be as follows:
1. Change the `scandir()` method to return an array of the found directories. The `delete_folders()` method then would just need to use the return value and would not need access to the property either.
    I imagine this was previously not done due to the person writing the code being uncomfortable with recursive functions, but it is perfectly possible to do so and would get rid of the unnecessary property.
    The method is currently a `void` method and sets the property.
2. Move both methods to the `Tests_Multisite_GetSpaceUsed` test class as clearly they are not needed elsewhere.

Even just implementing step 1 would already be a good improvement and would still leave these methods available to future and/or external integration tests.

I'm happy to convert this PR to the proper solution, but would like a second opinion first.

johnbillion reviewed (approved):
I think we need to be careful not to get distracted with refactoring otherwise functional code as we address dynamic properties, otherwise it could be a rabbit hole and slow things down.

Happy with this change 👍

jrfnl commented:
@johnbillion Agreed, but this _is_ still about the dynamic property as the change I propose in "step 1" removes the need for the (dynamic or otherwise) property altogether.

markjaquith reviewed (approved):
Agree that this code could be better, but this is the correct conservative fix for now. 👍🏻

SergeyBiryukov commented:
Thanks for the PR!

> Note: within the WP Core test suite, these methods are _only_ used in the `Tests_Multisite_GetSpaceUsed` test class.

Right, they were initially added for `wp_upload_dir()` tests in [r30658](https://core.trac.wordpress.org/changeset/30658), but subsequently removed from there in [r36565](https://core.trac.wordpress.org/changeset/36565).

> The "proper" solution, IMO, would be as follows:
> 
> 1. Change the `scandir()` method to return an array of the found directories. The `delete_folders()` method then would just need to use the return value and would not need access to the property either.
>    I imagine this was previously not done due to the person writing the code being uncomfortable with recursive functions, but it is perfectly possible to do so and would get rid of the unnecessary property.
>    The method is currently a `void` method and sets the property.

I think I would prefer this solution here, if possible. The `scandir()` method is clearly not intended for being used anywhere other than in `delete_folders()`, so we don't really need the property if `scandir()` can just return an array of the found directories.

jrfnl committed:
PHP 8.2 | WP_UnitTestCase_Base: minor refactor to remove need for $matched_dirs property

Follow-up to the earlier commit which explicitly declared the `$matched_dirs` property.

This commit implements the additional recommendation to remove the need for the - previously dynamic - property completely.

Effectively, this changes the `WP_UnitTestCase_Base::scandir()` method to return an array with matched directories instead of setting the property by using recursion in the method itself in an optimized manner.

Note the `array_merge()` not being in the loop itself, but at the very end of the function. This is for performance reasons.
See https://github.com/dseguy/clearPHP/blob/master/rules/no-array_merge-in-loop.md for a more detailed explanation of this.

I have verified in detail that the actual results of the previous version of this method and this version match, even when the paths passed are more complex and have deeper nested subdirectories.

jrfnl commented:
> I think I would prefer this solution here, if possible. The `scandir()` method is clearly not intended for being used anywhere other than in `delete_folders()`, so we don't really need the property if `scandir()` can just return an array of the found directories.

I've rebased the original commit (without changes) and added a second commit implementing the additional suggested changes per [1] for your review.

If approved, I'd suggest committing this to Core as one commit, but to allow for reviewing the difference between the solutions I've set this up as separate commits for this PR.

Please see the commit message for more details.

SergeyBiryukov commented:
Thanks for the PR! Merged in [r53937](https://core.trac.wordpress.org/changeset/53937).
